File: src/Scanner.Core/Checks/MissingTableHeaderCheck.cs
using System.Text.RegularExpressions;
using Scanner.Core.Rules;

namespace Scanner.Core.Checks;

/// <summary>
/// Checks table elements for missing header cells.
/// </summary>
public sealed class MissingTableHeaderCheck : ICheck
{
    private static readonly Regex TableRegex = new("<table(?<attrs>[^>]*)>(?<content>.*?)</table>", RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.Compiled);
    private static readonly Regex HeaderRegex = new("<th\\b", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <inheritdoc />
    public string Id => "missing-table-headers";

    /// <inheritdoc />
    public IReadOnlyCollection<string> ApplicableKinds { get; } = new[] { "html", "htm", "cshtml", "razor" };

    /// <inheritdoc />
    public IEnumerable<Issue> Run(CheckContext context, RuleDefinition rule)
    {
        var issues = new List<Issue>();

        foreach (Match match in TableRegex.Matches(context.Content))
        {
            var attrs = match.Groups["attrs"].Value;
            var role = AttributeParser.GetAttributeValue(attrs, "role");
            if (string.Equals(role, "presentation", StringComparison.OrdinalIgnoreCase)
                || string.Equals(role, "none", StringComparison.OrdinalIgnoreCase))
            {
                continue;
            }

            var content = match.Groups["content"].Value;
            if (HeaderRegex.IsMatch(content))
            {
                continue;
            }

            var line = TextUtilities.GetLineNumber(context.Content, match.Index);
            issues.Add(new Issue(rule.Id, Id, context.FilePath, line, "Table missing header cells.", match.Value));
        }

        return issues;
    }
}

File: node_engine/src/checks/MissingTableHeaderCheck.js
const { getAttributeValue } = require("./AttributeParser");
const { getLineNumber } = require("./TextUtilities");

const tableRegex = /<table(?<attrs>[^>]*)>(?<content>.*?)<\/table>/gis;
const headerRegex = /<th\b/i;

const MissingTableHeaderCheck = {
  id: "missing-table-headers",
  applicableKinds: ["html", "htm", "cshtml", "razor"],
  run(context, rule) {
    const issues = [];
    if (context.document?.querySelectorAll) {
      let searchStart = 0;
      for (const table of context.document.querySelectorAll("table")) {
        const role = table.getAttribute("role");
        const normalizedRole = role ? role.toLowerCase() : "";
        if (["presentation", "none"].includes(normalizedRole)) {
          continue;
        }

        if (table.querySelector("th")) {
          continue;
        }

        const evidence = table.outerHTML;
        const matchIndex = context.content.indexOf(evidence, searchStart);
        issues.push({
          ruleId: rule.id,
          checkId: MissingTableHeaderCheck.id,
          filePath: context.filePath,
          line: getLineNumber(context.content, matchIndex >= 0 ? matchIndex : searchStart),
          message: "Table missing header cells.",
          evidence
        });
        if (matchIndex >= 0) {
          searchStart = matchIndex + evidence.length;
        }
      }

      return issues;
    }

    for (const match of context.content.matchAll(tableRegex)) {
      const attrs = match.groups.attrs;
      const role = getAttributeValue(attrs, "role");
      const normalizedRole = role ? role.toLowerCase() : "";
      if (["presentation", "none"].includes(normalizedRole)) {
        continue;
      }

      const content = match.groups.content;
      if (headerRegex.test(content)) {
        continue;
      }

      const line = getLineNumber(context.content, match.index);
      issues.push({
        ruleId: rule.id,
        checkId: MissingTableHeaderCheck.id,
        filePath: context.filePath,
        line,
        message: "Table missing header cells.",
        evidence: match[0]
      });
    }

    return issues;
  }
};

module.exports = { MissingTableHeaderCheck };
